### PR TITLE
Task Center updates

### PR DESCRIPTION
**Pull request by cpeel**
The main focus for this work was to make task relationships bi-directional. If we say task B is related to task A, both tasks should show that relationship.

While I was here I addressed a few other pet peeves of mine, most notably that URLs inside comments and details weren't actually links.

You can demo this branch in the [taskcenter-related-tasks](https://www.pgdp.org/~cpeel/c.branch/taskcenter-related-tasks/tasks.php?action=show&task_id=2211) sandbox.

I'm also going to call [Task 33](https://www.pgdp.net/c/tasks.php?action=show&task_id=33) closed when this merges.

### CODE REVIEW

File: tasks.php [hunk 1]
```diff
@@ -1602,20 +1597,16 @@ function NotificationMail($tid, $message, $new_task = false)
...
-    $related_tasks = $row["related_tasks"];
     echo "<table class='tasks'>\n";
     echo "<tr><td style='width: 100%'><b>Related Tasks&nbsp;&nbsp;</b>";
     echo "<form action='$tasks_url' method='post'>";
     echo "<input type='hidden' name='action' value='add_related_task'>";
     echo "<input type='hidden' name='task_id' value='$tid'>";
     echo "<input type='number' name='related_task' min='1' required>&nbsp;&nbsp;";
     echo "<input type='submit' value='Add'>\n";
-    echo " (Add the number of an existing, related task. This is optional.)";
+    echo " (Add the number of an existing, related task.)";
```
Comment (srjfoo) [inline tasks.php]: Removing "optional" concerns me -- we don't want to imply that a related task is required.

File: tasks.php [hunk 2]
```diff
@@ -1916,4 +1975,13 @@ function title_string_for_task($pre_task)
     return sprintf("Task #%d: %s", $pre_task->task_id, $pre_task->task_summary);
 }
 
+// Convert any URLs into a clickable link in a string
+function make_urls_links($string)
+{
+    // from https://stackoverflow.com/questions/1960461/convert-plain-text-urls-into-html-hyperlinks-in-php
+    $url = '/(http|https|ftp|ftps)\:\/\/[a-zA-Z0-9\-\.]+\.[a-zA-Z]{2,3}(\/\S*)?/';
+    $string = preg_replace($url, '<a href="$0" target="_blank">$0</a>', $string);
```
Comment (srjfoo) [inline tasks.php]: Should we be using `target="_blank"` here?

File: tasks.php [hunk 3]
```diff
@@ -1645,20 +1636,88 @@ function RelatedTasks($tid)
...
     echo "<table class='tasks'>\n";
-    echo "<tr><td style='width: 100%'><b>Related Topic ID&nbsp;&nbsp;</b>";
+    echo "<tr><td style='width: 100%'><b>Related Topic</b>";
     echo "<form action='$tasks_url' method='post'>";
     echo "<input type='hidden' name='action' value='add_related_topic'>";
     echo "<input type='hidden' name='task_id' value='$tid'>";
     echo "<input type='number' name='related_posting' min='1' required>&nbsp;&nbsp;";
     echo "<input type='submit' value='Add'>\n";
-    echo " (Optional)";
+    echo " (Add the number of a forum topic.)";
```
[the remark this answers is not in the document]
Comment (cpeel) [reply tasks.php]: I don't think "optional" makes sense here. These fields aren't shown with creating or editing the task, just when viewing them. Someone would need to be viewing the page and then feel compelled to submit this form.

File: tasks.php (hunk 2 above)
Comment (cpeel) [reply tasks.php]: Nope, I will remove it.

File: tasks.php [hunk 4]
```diff
@@ -1058,6 +1047,11 @@ function hideSpan(id) { document.getElementById(id).style.display="none"; }
     echo "</table></form><br>\n";
 }
 
+function encode_array($a)
+{
```
[the remark this answers is not in the document]
Comment (cpeel) [reply tasks.php]: Yup, good call.

File: tasks.php [hunk 5]
```diff
@@ -1058,6 +1047,11 @@ function hideSpan(id) { document.getElementById(id).style.display="none"; }
...
+    return base64_encode(serialize($a));
+}
+
 function decode_array($str)
 {
     // Decode and deserialize a base64 encoded serialised array. Usually a
```
[the remark this answers is not in the document]
Comment (cpeel) [reply tasks.php]: Fixed

File: SETUP/db_schema.sql [hunk 6]
```diff
@@ -535,6 +535,21 @@ CREATE TABLE `tasks_comments` (
...
+#
+
+CREATE TABLE `tasks_related_tasks` (
+  `task_id_1` mediumint(9) NOT NULL,
+  `task_id_2` mediumint(9) NOT NULL,
+  PRIMARY KEY (`task_id_1`,`task_id_2`),
+  KEY `task_id_2` (`task_id_2`)
+) ENGINE=InnoDB DEFAULT CHARSET=latin1;
+# --------------------------------------------------------
+
```
[the remark this answers is not in the document]
Comment (cpeel) [reply SETUP/db_schema.sql]: That's a good question. I don't see a timestamp being all that useful here.

File: tasks.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (srjfoo) [reply tasks.php]: Good point -- I was not considering the context.

**Pull request merged**